libraries/xcode.rb

```diff
@@ -1,24 +1,54 @@
 include Chef::Mixin::ShellOut
 
 module MacOS
-  module Xcode
-    class << self
-      def installed?(semantic_version)
-        xcversion_output = shell_out(XCVersion.installed_xcodes).stdout.split
-        installed_xcodes = xcversion_output.values_at(*xcversion_output.each_index.select(&:even?))
-        installed_xcodes.include?(semantic_version)
-      end
+  class Xcode
+    attr_reader :version
+    attr_reader :credentials
 
-      def find_apple_id(data_bag_retrieval, node_credential_attributes)
-        if node_credential_attributes
-          { 'apple_id' => node_credential_attributes['user'],
-            'password' => node_credential_attributes['password'] }
-        else
-          data_bag_retrieval.call
-        end
-      rescue Net::HTTPServerException
-        Chef::Application.fatal!('No developer credentials supplied!')
+    def initialize(semantic_version, data_bag_retrieval = nil, node_credential_attributes = nil)
+      @semantic_version = semantic_version
+      developer_id = find_apple_id(data_bag_retrieval, node_credential_attributes)
+      @credentials = {
+        XCODE_INSTALL_USER:     developer_id['apple_id'],
+        XCODE_INSTALL_PASSWORD: developer_id['password'],
+      }
+      authenticate_with_apple(@credentials)
+      @version = apple_pseudosemantic_version(semantic_version)
+    end
+
+    def apple_pseudosemantic_version(semantic_version)
+      split_version = semantic_version.split('.')
```

Comment:
Have we considered using `Gem::Version` objects? They're much easier to inspect and compare than strings.

Reply:
I looked into trying this since I use `Gem` objects in the `Simulator` class. The problem I'm trying to solve is that Apple's versioning for new major releases of Xcode breaks semantic rules. The major releases have to be referred to as `Xcode 9`, not `Xcode 9.0`.  This method was built to strip the `.0` that users of the Xcode resource provide in the `version` property - I'm not sure how I can effectively utilize a `Version` object. 

Thanks for pointing out that it grew past it's original purpose though, I've split this into two methods.

Reply:
Sounds like we need our own version object. Would this work?

```ruby
2.5.1 (main):0 > class XcodeVersion < Gem::Version
2.5.1 (main):0 *   def major
2.5.1 (main):0 *     _segments.first
2.5.1 (main):0 *   end
2.5.1 (main):0 *   def minor
2.5.1 (main):0 *     _segments[1] || 0
2.5.1 (main):0 *   end
2.5.1 (main):0 *   def patch
2.5.1 (main):0 *     _segments[2] || 0
2.5.1 (main):0 *   end
2.5.1 (main):0 * end
=> :patch
2.5.1 (main):0 > XcodeVersion.create '9'
=> #<XcodeVersion "9">
2.5.1 (main):0 > v = _
=> #<XcodeVersion "9">
2.5.1 (main):0 > v.major
=> 9
2.5.1 (main):0 > v.minor
=> 0
2.5.1 (main):0 > v.patch
=> 0
```